tests/trainer/optimization/test_optimizers.py

```diff
@@ -661,3 +662,97 @@ def on_save_checkpoint(self, checkpoint):
     model.training_epoch_end = None
     trainer.fit(model)
     assert model.on_save_checkpoint_called
+
+
+def test_lr_scheduler_step_hook(tmpdir):
+    """Test that custom lr_schedulers works and `lr_scheduler_hook` is called at appropriate time."""
+
+    class CustomEpochScheduler:
+        def __init__(self, optimizer):
+            self.optimizer = optimizer
+
+        def step(self, epoch):
+            for param_group in self.optimizer.param_groups:
+                param_group["lr"] = param_group["lr"] / (epoch + 1)
+
+        def state_dict(self):
+            return {key: value for key, value in self.__dict__.items() if key != "optimizer"}
+
+        def load_state_dict(self, state_dict):
+            self.__dict__.update(state_dict)
+
+    class CustomBoringModel(BoringModel):
+        def __init__(self, learning_rate):
+            super().__init__()
+            self.learning_rate = learning_rate
+            self.layer1 = torch.nn.Linear(32, 2)
+            self.layer2 = torch.nn.Linear(32, 2)
+
+        def training_step(self, batch, batch_idx, optimizer_idx):
+            if optimizer_idx == 0:
+                output = self.layer1(batch)
+            else:
+                output = self.layer2(batch)
+
+            return self.loss(batch, output)
+
+        def training_epoch_end(self, *args, **kwargs):
+            pass
+
+        def lr_scheduler_step(self, scheduler, optimizer_idx, metric=None):
+            # step-level
+            if optimizer_idx == 0:
+                super().lr_scheduler_step(scheduler, optimizer_idx, metric)
+            # epoch-level
+            elif optimizer_idx == 1:
+                scheduler.step(epoch=self.current_epoch)
+
+        def configure_optimizers(self):
+            opt1 = torch.optim.SGD(self.layer1.parameters(), lr=self.learning_rate)
+            lr_scheduler1 = {"scheduler": torch.optim.lr_scheduler.StepLR(opt1, step_size=1), "interval": "step"}
+            opt2 = torch.optim.SGD(self.layer2.parameters(), lr=self.learning_rate)
+            lr_scheduler2 = CustomEpochScheduler(opt2)
+            return {"optimizer": opt1, "lr_scheduler": lr_scheduler1}, {
+                "optimizer": opt2,
+                "lr_scheduler": lr_scheduler2,
+            }
+
+    lr = 1e-2
+    max_epochs = 3
+    model = CustomBoringModel(learning_rate=lr)
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        enable_checkpointing=False,
+        logger=False,
+        max_epochs=max_epochs,
+        limit_train_batches=2,
+        limit_val_batches=0,
+    )
+    trainer.fit(model)
+
+    for param_group in trainer.optimizers[1].param_groups:
+        assert param_group["lr"] == lr / math.factorial(max_epochs)
+
+
+def test_invalid_lr_scheduler(tmpdir):
+    """Test that custom lr_schedulers works and `lr_scheduler_hook` is called at appropriate time."""
```

Comment:
the docstring does not express what is being tested

Reply:
my bad!